File: lib/pocs.go
package lib

import (
	"fmt"
	"io/ioutil"

	"gopkg.in/yaml.v3"
)

type Poc struct {
	Name       string `yaml:"name"`
	Set        string `yaml:"set"`
	Rules      Rules  `yaml:"rules"`
	Expression string `yaml:"expression"`
	Detail     Detail `yaml:"detail"`
}

type Rules []RuleItem

type RuleItem struct {
	key  string
	rule Rule
}

func (r *Rules) UnmarshalYAML(unmarshal func(interface{}) error) error {

	var tmp = make(map[string]Rule)
	if err := unmarshal(&tmp); err != nil {
		return err
	}

	for key, value := range tmp {
		*r = append(*r, RuleItem{key, value})
	}
	return nil
}

type Rule struct {
	Request    Request `yaml:"request"`
	Expression string  `yaml:"expression"`
}

type Request struct {
	Method  string            `yaml:"method"`
	Path    string            `yaml:"path"`
	Headers map[string]string `yaml:"headers"`
	Body    string            `yaml:"body"`
}

type Detail struct {
	Author string   `yaml:"author"`
	Links  []string `yaml:"links"`
}

func LoadPoc(fileName string) (*Poc, error) {
	p := &Poc{}

	yamlFile, err := ioutil.ReadFile(fileName)
	if err != nil {
		fmt.Println(err)
		return nil, err
	}
	err = yaml.Unmarshal(yamlFile, p)

	return p, err
}

File: src/main/java/com/github/trganda/pocs/Pocs.java
package com.github.trganda.pocs;

/**
 * name: poc-yaml-example-com
 * transport: http
 * set:
 *     r1: randomInt(800000000, 1000000000)
 *     r2: randomInt(800000000, 1000000000)
 *     r3: base64(b'asdf')
 * rules:
 *     r1:
 *         request:
 *             method: POST
 *             path: "/index.jsp?c={{r3}}"
 *             body: "a=b&b=a"
 *         expression: |
 *             response.status==200 && response.body.bcontains(b'Example Domain')
 * expression: |
 *     r1()
 * detail:
 *     author: name(link)
 *     links:
 *         - http://example.com
 */
public class Pocs {
    public String name;
    public String transport;
    public String expression;
    public Sets set;
    public Rules rules;
    public Detail detail;

    public Pocs() {

    }

    public Pocs(String name, String transport, String expression, Detail detail) {
        this.name = name;
        this.transport = transport;
        this.expression = expression;
        this.detail = detail;
    }

    @Override
    public String toString() {
        return "Pocs{" +
                "name='" + name + '\'' +
                ", transport='" + transport + '\'' +
                ", expression='" + expression + '\'' +
                ", set=" + set +
                ", rules=" + rules +
                ", detail=" + detail +
                '}';
    }
}
